src/vs/editor/contrib/stickyScroll/browser/stickyScroll.ts

```diff
@@ -0,0 +1,338 @@
+			if (token.isCancellationRequested) {
+				return;
+			}
+			this._ranges = [];
+			for (const outline of outlineModel.children.values()) {
+				if (outline instanceof OutlineElement) {
+					this._findLineRanges(outline, 1);
+				}
+			}
+			this._ranges = this._ranges.sort(function (a, b) {
+				if (a[0] !== b[0]) {
+					return a[0] - b[0];
+				} else if (a[1] !== b[1]) {
+					return b[1] - a[1];
+				} else {
+					return a[2] - b[2];
+				}
+			});
+			let previous: number[] = [];
+			for (const [index, arr] of this._ranges.entries()) {
+				const [start, end, _depth] = arr;
+				if (previous[0] === start && previous[1] === end) {
+					this._ranges.splice(index, 1);
+				} else {
+					previous = arr;
+				}
+			}
+		}
+	}
+
+	private _renderStickyScroll() {
+
+		if (!(this._editor.hasModel())) {
+			return;
+		}
+
+		const lineHeight = this._editor.getOption(EditorOption.lineHeight);
+		const model = this._editor.getModel();
+
+		const scrollTop = this._editor.getScrollTop();
+		let scrollDirection: ScrollDirection;
+		if (this._lastScrollPosition < scrollTop) {
+			scrollDirection = ScrollDirection.Down;
+		} else {
+			scrollDirection = ScrollDirection.Up;
+		}
+		this._lastScrollPosition = scrollTop;
+
+		const scrollTopFromLengthOfArray: number = this._editor.getScrollTop() + this.stickyScrollWidget.codeLineCount * lineHeight + 1;
+		this.stickyScrollWidget.emptyRootNode();
+		const beginningLinesConsidered = new Set();
+
+		for (const [index, arr] of this._ranges.entries()) {
+			const [start, end, depth] = arr;
+			const scrollTopFromDepth: number = this._editor.getScrollTop() + depth * lineHeight + 1;
+
+			if (!beginningLinesConsidered.has(start)) {
+				if (this._editor.getScrollTop() + (depth - 1) * lineHeight + 1 >= (end - 1) * lineHeight && this._editor.getScrollTop() + (depth - 1) * lineHeight < end * lineHeight - 2) {
+					beginningLinesConsidered.add(start);
+					this.stickyScrollWidget.pushCodeLine(new StickyScrollCodeLine(model.getLineContent(start), start, this._editor, -1, (depth - 1) * lineHeight + end * lineHeight - this._editor.getScrollTop() - depth * lineHeight));
+					break;
+				}
+				else if (scrollDirection === ScrollDirection.Down && scrollTopFromDepth > start * lineHeight && scrollTopFromDepth < end * lineHeight) {
+					beginningLinesConsidered.add(start);
+					this.stickyScrollWidget.pushCodeLine(new StickyScrollCodeLine(model.getLineContent(start), start, this._editor, 0));
+
+				} else if (scrollDirection === ScrollDirection.Up && scrollTopFromLengthOfArray > start * lineHeight && scrollTopFromLengthOfArray < end * lineHeight ||
+					scrollDirection === ScrollDirection.Up && scrollTopFromDepth >= start * lineHeight && scrollTopFromDepth < (end - 1) * lineHeight) {
+					beginningLinesConsidered.add(start);
+					this.stickyScrollWidget.pushCodeLine(new StickyScrollCodeLine(model.getLineContent(start), start, this._editor, 0));
+				}
+			} else {
+				this._ranges.splice(index, 1);
```

Comment:
It seems to me that it would be better if updating `this._ranges` would be limited to the `_updateOutlineModel` method and not in the rendering code.